electrum/onion_message.py

```diff
@@ -0,0 +1,759 @@
+            next_peer = self.lnwallet.peers.get(next_node_id)
+            if not next_peer:
+                self.logger.info(f'next node {next_node_id.hex()} not a peer, dropping message')
+                return
+
+        # blinding override?
+        next_blinding_override = recipient_data.get('next_blinding_override')
+        if next_blinding_override:
+            next_blinding = next_blinding_override.get('blinding')
+        else:
+            # E_i+1=SHA256(E_i||ss_i) * E_i
+            blinding_factor = sha256(blinding + shared_secret)
+            blinding_factor_int = int.from_bytes(blinding_factor, byteorder="big")
+            next_public_key_int = ecc.ECPubkey(blinding) * blinding_factor_int
+            next_blinding = next_public_key_int.get_public_key_bytes()
+
+        if is_dummy_hop:
+            self.process_onion_message_packet(next_blinding, onion_packet)
+            return
+
+        self.submit_forward(onion_packet=onion_packet, blinding=next_blinding, node_id=next_node_id)
+
+    def on_onion_message(self, payload: dict) -> None:
+        """handle arriving onion_message."""
+        blinding = payload.get('blinding')
+        if not blinding:
+            self.logger.error('missing blinding')
+            return
+        packet = payload.get('onion_message_packet')
+        if payload.get('len', 0) != len(packet):
+            self.logger.error('invalid/missing length')
+            return
+
+        self.logger.debug('handling onion message')
+
+        onion_packet = OnionPacket.from_bytes(packet)
+        self.process_onion_message_packet(blinding, onion_packet)
+
+    def process_onion_message_packet(self, blinding: bytes, onion_packet: OnionPacket) -> None:
+        our_privkey = blinding_privkey(self.lnwallet.node_keypair.privkey, blinding)
+        processed_onion_packet = process_onion_packet(onion_packet, our_privkey, tlv_stream_name='onionmsg_tlv')
+        payload = processed_onion_packet.hop_data.payload
+
+        self.logger.debug(f'onion peeled: {processed_onion_packet!r}')
+
+        if not processed_onion_packet.are_we_final:
+            if any([x not in ['encrypted_recipient_data'] for x in payload.keys()]):
+                self.logger.error('unexpected data in payload')  # non-final nodes only encrypted_recipient_data
+                return
+
+        # decrypt
+        shared_secret = get_ecdh(self.lnwallet.node_keypair.privkey, blinding)
+        recipient_data = decrypt_onionmsg_data_tlv(
+            shared_secret=shared_secret,
+            encrypted_recipient_data=payload['encrypted_recipient_data']['encrypted_recipient_data']
+        )
+
+        self.logger.debug(f'parsed recipient_data: {recipient_data!r}')
+
+        if processed_onion_packet.are_we_final:
+            self.on_onion_message_received(recipient_data, payload)
+        else:
+            self.on_onion_message_forward(recipient_data, processed_onion_packet.next_packet, blinding, shared_secret)
```

Comment:
When we are just a light client, e.g. running on a phone, do we want to forward messages?
Payment forwarding is gated behind a config variable. Should this also be?

Reply:
Although it would be unlikely that a mobile electrum will be included in a blinded path, we should probably do it for consistency :)

Reply:
I was thinking about attack surface reduction.
Anyway, if you agree there is no point for a mobile node to forward, I think we could potentially reuse `config.EXPERIMENTAL_LN_FORWARD_PAYMENTS` for this (instead of adding a new config var).